stdlib/public/core/Rotate.swift

```diff
@@ -0,0 +1,716 @@
+                case (true, true):
+                    // The partitions have the same length.
+                    return middle
+                }
+            }
+        }
+    }
+
+    /// Rotates the elements of the collection such that the value at the given
+    /// index moves rightward to just before `endIndex`.
+    ///
+    /// Passing `endIndex` or its immediate predecessor (if any) as `middle` has
+    /// no effect.
+    ///
+    /// The following example rotates the elements of an array of characters so
+    /// the second character moves to be the last:
+    ///
+    ///     var café: [Character] = ["C", "a", "f", "é"]
+    ///     café.rotate(toLast: café.index(after: café.startIndex))
+    ///     print(café)
+    ///     // Prints "["f", "é", "C", "a"]"
+    ///
+    /// - Precondition: `middle` must be a valid index of this collection.
+    ///
+    /// - Parameter middle: The index of the element whose value will move to
+    ///   just before `endIndex`, unless it's `endIndex` (then it's ignored).
+    ///
+    /// - Returns: The new index of the value originally at `startIndex`.
+    ///
+    /// - Postcondition: The collection is a right-rotation of its old state,
+    ///   *i.e.* `newValue` equals `oldValue[middle>..] + oldValue[...middle]`.
+    ///
+    /// - Complexity: O(*n*), where *n* is the number of elements in the
+    ///   collection.
+    @inlinable
+    @discardableResult
+    public mutating func rotate(toLast middle: Index) -> Index {
+        return rotate(toFirst: index(middle, offsetBy: +1, limitedBy: endIndex)
+            ?? startIndex)
+    }
+}
+
+/// A collection that presents the elements of its base collection
+/// in rotated order.
+///
+/// - Note: This type is the result from either `x.rotated(toFirst:)` or
+///   `x.rotated(toLast:)` for a given collection `x`.
+///
+/// The `rotated` methods are always lazy, but do not implicitly confer
+/// laziness on algorithms applied to their results.  In other words, for
+/// ordinary collections `c`:
+///
+/// * `c.rotated(toFirst: i)` does not create new storage
+/// * `c.rotated(toFirst: i).map(f)` maps eagerly and returns a new array
+/// * `c.lazy.rotated(toFirst: i).map(f)` maps lazily and returns a
+///   `LazyMapCollection`
+/// * (The above similarly apply to `rotated(toLast:)`.)
+@_fixed_layout
+public struct RotatedCollection<Base: Collection> {
+    /// The underlying collection.
+    public let _base: Base
+    /// The underlying index for the rotated collection's first element.
+    @usableFromInline
+    internal var _origin: Base.Index
+
+    /// Creates an instance that presents the elements of `base` in a rotated
+    /// order with the given index as the new start-index.
+    ///
+    /// - Complexity: O(1)
+    @inlinable
+    internal init(_base: Base, asFirst middle: Base.Index) {
+        self._base = _base
+        _origin = middle < _base.endIndex ? middle : _base.startIndex
```

Comment:
Again, not convinced that out-of-bounds middle shouldn't be a precondition failure.